### PR TITLE
KEYCLOAK-18445 Add support for cross-site model tests

### PR DESCRIPTION
**Pull request by martin-kanis**

### CODE REVIEW

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java [hunk 1]
```diff
@@ -143,18 +147,67 @@ public void testUserSessionInitializerWithDeletingClient() {
...
+                    siteName.set(CONFIG.scope("connectionsInfinispan", "default").get("siteName", "site-1"));
+
+                    // create a user session in the first node
+                    UserSessionModel userSessionModel = withRealm(realmId, (session, realm) -> {
+                        final UserModel user = session.users().getUserByUsername(realm, "user1");
+                        return session.sessions().createUserSession(realm, user, "un1", "ip1", "auth", false, null, null);
+                    });
+                    userSessionId.set(userSessionModel.getId());
+                } else {
+                    if (siteName.get().equals(CONFIG.scope("connectionsInfinispan", "default").get("siteName", "site-1"))) {
```
Comment (mhajas) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:179]: This if seems to be redundant.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java [hunk 2]
```diff
@@ -157,7 +157,6 @@ public void testPersistenceWithLoad() {
             // Assert offline sessions
             RealmModel realm = session.realms().getRealm(realmId);
             List<UserSessionModel> loadedSessions = loadPersistedSessionsPaginated(session, true, 2, 2, 3);
-            assertSessions(loadedSessions, new String[] { origSessions[0].getId(), origSessions[1].getId(), origSessions[2].getId() });
```
Comment (mhajas) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java:160]: Should we check for size here?

**Review by mhajas — COMMENTED** [review]
Great work @martin-kanis. This will really help with HotRod testing! I added some comments/questions to the code.

File: model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java [hunk 3]
```diff
@@ -205,7 +196,8 @@ protected void initEmbedded() {
 
         if (clustered) {
             String jgroupsUdpMcastAddr = config.get("jgroupsUdpMcastAddr", System.getProperty(InfinispanConnectionProvider.JGROUPS_UDP_MCAST_ADDR));
-            configureTransport(gcb, topologyInfo.getMyNodeName(), topologyInfo.getMySiteName(), jgroupsUdpMcastAddr);
+            configureTransport(gcb, topologyInfo.getMyNodeName(), topologyInfo.getMySiteName(), jgroupsUdpMcastAddr,
+                    "default-configs/default-keycloak-jgroups-udp.xml", this.getClass().getClassLoader());
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java:200]: Good point. No removed it and used the classloader in `InfinispanUtil`.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java [hunk 4]
```diff
@@ -92,4 +113,142 @@ public static long toHotrodTimeMs(BasicCache ispnCache, long lifespanOrigMs) {
         return lifespanOrigMs;
     }
 
+    private static final Object CHANNEL_INIT_SYNCHRONIZER = new Object();
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java:116]: No, I just moved those methods from `DefaultInfinispanConnectionProviderFactory`.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/CrossDCInfinispan.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    private final HotRodServerRule hotRodServerRule = new HotRodServerRule();
+
+    private static final AtomicInteger NODE_COUNTER = new AtomicInteger();
+
+    private static final String SITE_1_MCAST_ADDR = "228.5.6.7";
+
+    private static final String SITE_2_MCAST_ADDR = "228.6.7.8";
+
+    private Object lock = new Object();
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/CrossDCInfinispan.java:40]: Added

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java [hunk 6]
```diff
@@ -143,18 +147,67 @@ public void testUserSessionInitializerWithDeletingClient() {
...
+        });
+
+        assertThat(containsSession.get(), everyItem(is(true)));
+    }
+
+    // Create sessions in persister + infinispan, but then delete them from infinispan cache by reinitializing keycloak session factory
+    private AtomicReferenceArray createSessionsInPersisterOnly() {
         UserSessionModel[] origSessions = inComittedTransaction(session -> { return UserSessionPersisterProviderTest.createSessions(session, realmId); });
-        String[] res = new String[origSessions.length];
+        AtomicReferenceArray<String> res = new AtomicReferenceArray(origSessions.length);
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:210]: Changed

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java [hunk 7]
```diff
@@ -143,18 +147,67 @@ public void testUserSessionInitializerWithDeletingClient() {
...
+
+                            containsSession.get().add(localSessions.containsKey(userSessionId.get()));
+                            containsSession.get().add(remoteSessions.containsKey(userSessionId.get()));
+                        });
+                    }
+                }
+            }
+        });
+
+        assertThat(containsSession.get(), everyItem(is(true)));
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:204]: Yeah I added a check for size and also little logic to distinguish runs with Hot Rod and without Hot Rod (final size depends on that).

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java (hunk 1 above)
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:179]: Good point. I simplified it.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java (hunk 2 above)
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java:160]: I returned this back. Not sure why I removed it.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java [hunk 8]
```diff
@@ -109,17 +116,16 @@ public void testUserSessionInitializer() {
...
+            assertSessionLoaded(loadedSessions, origSessionIds.get(0), session.users().getUserByUsername(realm, "user1"), "127.0.0.1", started, started, "test-app", "third-party");
+            assertSessionLoaded(loadedSessions, origSessionIds.get(1), session.users().getUserByUsername(realm, "user1"), "127.0.0.2", started, started, "test-app");
+            assertSessionLoaded(loadedSessions, origSessionIds.get(2), session.users().getUserByUsername(realm, "user2"), "127.0.0.3", started, started, "test-app");
         });
     }
 
     @Test
     public void testUserSessionInitializerWithDeletingClient() {
-        String[] origSessionIds = createSessionsInPersisterOnly();
+        AtomicReferenceArray<String> origSessionIds = createSessionsInPersisterOnly();
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionInitializerTest.java:128]: Yeah it's not needed. Reverted.

File: model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionAdapter.java [hunk 9]
```diff
@@ -111,22 +111,22 @@ public boolean isOffline() {
...
-                        AuthenticatedClientSessionModel clientSession = session.sessions()
-                                .getClientSession(this, client, entry.getValue(), isOffline());
-                        if (clientSession != null) {
-                            result.put(clientUUID, clientSession);
-                        }
-                    } else {
-                        removedClientUUIDS.add(clientUUID);
-                    }
-                });
+        // to avoid concurrentModificationException
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionAdapter.java:114]: You can reproduce the issue like this:

- revert changes in `MapUserSessionAdapter`
- For example from `UserSessionProviderModelTest#testExpiredClientSessions` in last `inComittedTransaction` call `userSession.getAuthenticatedClientSessions().entrySet().stream().forEach(c -> {});`
- `mvn test -f testsuite/model/pom.xml -Pmap -Dtest=UserSessionProviderModelTest#testExpiredClientSessions`

```
java.util.ConcurrentModificationException
	at java.base/java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:719)
	at java.base/java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:751)
	at java.base/java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:749)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.keycloak.models.map.userSession.MapUserSessionAdapter.getAuthenticatedClientSessions(MapUserSessionAdapter.java:116)
```
I'm not sure why it's happening.

File: services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java [hunk 10]
```diff
@@ -189,7 +189,7 @@ protected ExecutorService createPoolEmbedded(String taskType, KeycloakSession se
             // Same like Executors.newCachedThreadPool. Besides that "min" and "max" are configurable
             return new ThreadPoolExecutor(min, max,
                     60L, TimeUnit.SECONDS,
-                    new SynchronousQueue<Runnable>(),
+                    new ArrayBlockingQueue<>(1024),
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java]: I can't remember the exact error message but it was related to `InfinispanNotificationsManager` and some cluster listener. Maybe @hmlnarik can remember more?

Anyway, I reverted the change and I don't see the error again. Maybe it was present only in earlier phases of this PR.

**Comment by martin-kanis** [comment]
@mhajas Thank you for the feedback! I addressed your comments and answered your questions. 

CrossDC tests failed but I don't see how that could be related to this PR.

File: model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionAdapter.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionAdapter.java:114]: I found out the reason of the `CME`. The problem is, that `session.sessions().getClientSession()` [udpates](https://github.com/keycloak/keycloak/blob/master/model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java#L116) `entity.getAuthenticatedClientSessions()` map when the clientSession expires. 

We need to do the `getClientSession` call outside of looping through `authenticatedClientSessions`. I cannot come up with any nice solution, so the copy of `authenticatedClientSessions` is probably the best. But if you can find anything better it would be great.

**Review by mhajas — DISMISSED** [review]
Thank you for the fixes @martin-kanis! I am approving, but would be great to get some feedback from @hmlnarik regarding `ArrayBlockingQueue` vs `SynchronousQueue` [1] before merging.

[1] https://github.com/keycloak/keycloak/pull/8267#discussion_r709270460

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java [hunk 11]
```diff
@@ -594,8 +594,8 @@ public static void assertSession(UserSessionModel session, UserModel user, Strin
         assertEquals(user.getUsername(), session.getLoginUsername());
         assertEquals("form", session.getAuthMethod());
         assertTrue(session.isRememberMe());
-        assertTrue(session.getStarted() >= started - 1 && session.getStarted() <= started + 1);
-        assertTrue(session.getLastSessionRefresh() >= lastRefresh - 1 && session.getLastSessionRefresh() <= lastRefresh + 1);
+        assertTrue(session.getStarted() >= started - 2 && session.getStarted() <= started + 2);
+        assertTrue(session.getLastSessionRefresh() >= lastRefresh - 2 && session.getLastSessionRefresh() <= lastRefresh + 2);
```
Comment (hmlnarik) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java:598]: Why is this change necessary?

File: services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (martin-kanis) [reply services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java]: The problem showed up again after I reverted the change. Here is the tests [output].(https://pipelines.actions.githubusercontent.com/ENJd2H9lx5qREjW4so5u3q3ePn6aZgA0m90QfCJXr9DcNaynQj/_apis/pipelines/1/runs/492/signedlogcontent/21?urlExpires=2021-10-04T10%3A17%3A43.7621894Z&urlSigningMethod=HMACV1&urlSignature=MlpIVVmvvG3vLpPdAMrKxmjptOtPsYXWmW%2B2KtyyGwA%3D)

File: testsuite/model/src/main/java/org/keycloak/testsuite/model/KeycloakModelParameters.java [hunk 12]
```diff
@@ -67,4 +67,11 @@ public Statement instanceRule(Statement base, Description description) {
         return base;
     }
 
+    public void suiteRule(Config cf) {
+
+    }
+
+    public void afterClass() {
+
+    }
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply testsuite/model/src/main/java/org/keycloak/testsuite/model/KeycloakModelParameters.java]: I changed `suiteRule` to `beforeSuite` and removed `afterClass` and used `classRule`.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java [hunk 13]
```diff
@@ -37,6 +56,8 @@
  */
 public class InfinispanUtil {
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java]: moved

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java [hunk 14]
```diff
@@ -92,4 +113,142 @@ public static long toHotrodTimeMs(BasicCache ispnCache, long lifespanOrigMs) {
         return lifespanOrigMs;
     }
 
+    private static final Object CHANNEL_INIT_SYNCHRONIZER = new Object();
+
+    public static void configureTransport(GlobalConfigurationBuilder gcb, String nodeName, String siteName, String jgroupsUdpMcastAddr,
+                                      String jgroupsUdpConfigPath) {
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/InfinispanUtil.java]: changed

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java [hunk 15]
```diff
@@ -400,6 +406,13 @@ public void cleanEnvironment() {
         KeycloakModelUtils.runJobInTransaction(getFactory(), this::cleanEnvironment);
     }
 
+    @AfterClass
+    public static void freeResources() {
+        for (KeycloakModelParameters kmp : KeycloakModelTest.MODEL_PARAMETERS) {
+            kmp.afterClass();
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java]: done

File: testsuite/model/src/main/resources/hotRod/hotRod1.xml [hunk 16]
```diff
@@ -0,0 +1,43 @@
+<infinispan>
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply testsuite/model/src/main/resources/hotRod/hotRod1.xml]: renamed

File: testsuite/model/src/main/java/org/keycloak/testsuite/model/HotRodServerRule.java [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    private void getCaches(String ...cache) {
+        for (String c: cache) {
+            hotRodCacheManager.getCache(c, true);
+            hotRodCacheManager2.getCache(c, true);
+        }
+    }
+
+    private void createKeycloakCaches(boolean async, String ...cache) {
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply testsuite/model/src/main/java/org/keycloak/testsuite/model/HotRodServerRule.java]: fixed

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java (hunk 11 above)
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java]: It started to fail after I changed `userSessionprovider.removeLocalUserSessions()` to `reinitializeKeycloakSessionFactory` in `createSessionsInPersisterOnly()`. 
At the end I decided to fix it more nicely by capturing `started` directly from the created offline session rather than from `Time.currentTime()` after the `createSessionsInPersisterOnly() is called`.

File: services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java:192]: > Why is this change necessary?
> 
> If I understand correctly, this means, that when 1024 tasks are waiting for execution a `RejectedExecutionException` exception is thrown. Should we catch this exception for example here: https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/saml/SamlService.java#L387?

No, there are many other places that use this functionality, and the exhausted pool should be visible to the administrator to take appropriate action in configuring the pool. FTR, the default queue in Wildfly is unbounded.

**Review by hmlnarik — DISMISSED** [review]
Thank you for the PR!

**Review by hmlnarik — APPROVED** [review]
